### DISCUSSION TITLE
Manual management of letsencrypt certificate?

### DISCUSSION THREAD

**Discussion by djnicholson** [nextcloud/all-in-one#3982] [Questions] [answered]
I have a Nextcloud AIO instance that is not behind a reverse proxy, was exposed to the Internet when initially setup but is now no longer exposed to the Internet.  The domain name for the instance is a valid Internet domain name that can be resolved in public DNS (but resolves to a private IP only routable when connected to my VPN).

I am assuming that Nextcloud will fail to renew its letsencrypt certificate when it is close to expiry.  Is it possible for me to configure Nextcloud to use DNS ACME challenges instead of HTTP ACME challenges?  If I can't do that, is it possible for me to manage the certificate manually (perhaps I could write my own script to maintain a valid letsencrypt cert, and mount the location containing the certificate and keypair so that it is accessible to the Nextcloud container?)

**Comment by szaimen** [comment, marked as answer]
> I am assuming that Nextcloud will fail to renew its letsencrypt certificate when it is close to expiry. 

indeed

> Is it possible for me to configure Nextcloud to use DNS ACME challenges instead of HTTP ACME challenges? 

See https://github.com/nextcloud/all-in-one/blob/main/local-instance.md

> If I can't do that, is it possible for me to manage the certificate manually (perhaps I could write my own script to maintain a valid letsencrypt cert, and mount the location containing the certificate and keypair so that it is accessible to the Nextcloud container?)

This is not possible.

**Comment by djnicholson** [comment]
Thanks, I was able to get this working using nginx as a reverse proxy. I used the `jonasal/nginx-certbot` Docker image, run as follows:

```
docker run -d \
  --name=nginx \
  --restart unless-stopped \
  -p 443:443 \
  --env CERTBOT_EMAIL=$EMAIL \
  --env CERTBOT_AUTHENTICATOR=dns-route53 \
  --env AWS_ACCESS_KEY_ID=$AWS_KEY_ID \
  --env AWS_SECRET_ACCESS_KEY=$AWS_KEY \
  -v $DATA/nginx/secrets:/etc/letsencrypt \
  -v $DATA/nginx/config:/etc/nginx/user_conf.d:ro \
  jonasal/nginx-certbot:latest
```

I used the following nginx config:

```
map $http_upgrade $connection_upgrade {
    default upgrade;
    '' close;
}

server {
    # Listen to port 443 on IPv4
    listen 443 ssl default_server reuseport;

    http2 on;
    http3 on;
    quic_retry on;
    add_header Alt-Svc 'h3=":443"; ma=86400';
    listen 443 quic reuseport;

    # Domain names this server should respond to.
    server_name mydomain.net;

    # Load the certificate files.
    ssl_certificate         /etc/letsencrypt/live/mydomain.net/fullchain.pem;
    ssl_certificate_key     /etc/letsencrypt/live/mydomain.net/privkey.pem;
    ssl_trusted_certificate /etc/letsencrypt/live/mydomain.net/chain.pem;

    # Load the Diffie-Hellman parameter.
    ssl_dhparam /etc/letsencrypt/dhparams/dhparam.pem;

    location / {
        proxy_pass http://1.2.3.4:11000$request_uri;

        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Port $server_port;
        proxy_set_header X-Forwarded-Scheme $scheme;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header Accept-Encoding "";
        proxy_set_header Host $host;

        client_body_buffer_size 512k;
        proxy_read_timeout 86400s;
        client_max_body_size 0;

        # Websocket
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;
    }

}
```

This does seem a bit overkill though.

I assume that one of the Nextcloud containers runs certbot.  Would it be possible to allow people to configure it to use a different certbot authenticator (e.g. dns-route53) by setting an environment variable in the mastercontainer?
